windows-agent/internal/daemon/networking_mock.go

```diff
@@ -1,18 +1,210 @@
+	case ok:
+		m.adapters = append(adaptersList,
+			mockIPAdapterAddresses{ipconfig: m, name: "Ethernet adapter vEthernet (Default Switch)", desc: " Hyper-V Virtual Ethernet Adapter", ip: net.IPv4(172, 27, 48, 1)},
+			mockIPAdapterAddresses{ipconfig: m, name: "Ethernet adapter vEthernet (WSL (Hyper-V firewall))", desc: " Hyper-V Virtual Ethernet Adapter #2", ip: localIP})
+	}
+
+	return &m.adapters[0], nil
+}
+
+// getLocalPrivateIP returns one non loopback local private IP of the host.
+func getLocalPrivateIP() net.IP {
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return nil
+	}
+	for _, addr := range addrs {
+		if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() && ipnet.IP.IsPrivate() {
+			return ipnet.IP
+		}
+	}
+	return nil
+}
+func (m *mockIPAdapterAddresses) Next() ipAdapterAddresses {
+	return m.ipconfig.next()
+}
+
+func (m *mockIPAdapterAddresses) Description() string {
+	return m.desc
+}
+
+func (m *mockIPAdapterAddresses) IP() net.IP {
+	return m.ip
+}
+
+func (m *mockIPAdapterAddresses) FriendlyName() string {
+	return m.name
+}
+
+func (m *mockIPConfig) next() ipAdapterAddresses {
+	if m.current+1 >= len(m.adapters) {
+		return nil
+	}
+	m.current++
+	return &m.adapters[m.current]
+}
+
+type mockIPConfig struct {
+	state    mockIPAdaptersState
+	adapters []mockIPAdapterAddresses
+	current  int
+}
+type mockIPAdapterAddresses struct {
+	ipconfig   *mockIPConfig
+	name, desc string
+	ip         net.IP
+}
+
+type mockWslSystem struct {
+	netmode  string
+	extraEnv []string
+	cmdError bool
+}
+
+func newWslSystemMock(netmode string, extraEnv []string, cmdError bool) *mockWslSystem {
+	return &mockWslSystem{netmode: netmode, extraEnv: extraEnv, cmdError: cmdError}
+}
+
+func (m *mockWslSystem) Command(ctx context.Context, name string, args ...string) *exec.Cmd {
+	if !testing.Testing() {
+		panic("mockWslSystem can only be used within a test")
+	}
+
+	goArgs := append([]string{"test", "-run", "^TestWithWslSystemMock$", "--", name}, args...)
+	// Switches
+	env := append(os.Environ(), m.extraEnv...)
+	env = append(env,
+		fmt.Sprintf("%s=1", "UP4W_MOCK_EXECUTABLE"),
+		fmt.Sprintf("%s=%s", "UP4W_MOCK_NETWORKING_MODE", m.netmode),
+	)
+	if m.cmdError {
+		env = append(env, fmt.Sprintf("%s=1", "UP4W_MOCK_NETWORKING_MODE_ERROR"))
+	}
+
+	//nolint: gosec // Subprocess launched with variable (gosec) intentionally so we can mock it.
+	c := exec.CommandContext(ctx, "go", goArgs...)
+	c.Env = env
+	return c
+}
+
+// WslSystemMock mocks commands running inside the WSL system distro.
+// Add it to your package_test with:
+//
+//	func TestWithWslSystemMock(t *testing.T) { daemon.WslSystemMock(t) }
+//
+//nolint:thelper // This is a faux test used to mock commands running via `wsl -- system`
+func WslSystemMock(t *testing.T) {
+	// Setup
+	if t.Name() != "TestWithWslSystemMock" {
+		panic("The WslSystemMock faux test must be named TestWithWslSystemMock")
```

Comment:
codecov is complaining that my mock of WSL is not covered, which was created solely for testing and is broadly used. How should we handle that? Should the command invoking the test mock be `go test -coverpkg....` matching the same directory as the main tests are running? 🤷🏽‍♂️ 

I also considered instead of running `go test` as the executable mock, run the same binary the test process is running. Since the function implementing the executable is part of the same test file, `go test` already compiled it when running the test cases, so `os.Executable()` is a valid target for the mock, with a minor adjustment to the command line invocation: instead of `go test -run ^TESTNAME$` it would be `thisbinary -test.run ^TESTNAME$`. I suspect that would speed up tests with mocks, but I'm not quite sure how much, as `go test` probably runs part of the compiler toolchain at least to figure out that there is no need to compile anything.

Reply:
This is not really how we do it. I think the issues comes from the way the you are running the subprocess.

I suggest that instead of passing this, you are passing a subcommand path (with args) to the real object, a little bit like https://github.com/ubuntu/authd/blob/main/internal/users/localgroups/localgroups.go#L19 (but attached on the struct).
In the tests, then, you pass another command to the constructor:  https://github.com/ubuntu/authd/blob/main/internal/users/localgroups/localgroups_test.go#L213C4-L213C14.

This command reference another test from the same function:
https://github.com/ubuntu/authd/blob/main/internal/users/localgroups/localgroups_test.go#L257. In most of the project, if you don’t need it for special build or integration tests, the whole implementation is there, like in https://github.com/ubuntu/adsys/blob/main/internal/policies/apparmor/apparmor_test.go#L212.

You can see in that case that this is referenced to a subpackage, which is the "mock", that will be excluded from codecov (manually or not): https://github.com/ubuntu/authd/blob/main/internal/users/localgroups/tests/tests.go (with the according assertion).

Note that contrary to this case, I would advise to ensure this is running on tests, like  with https://github.com/ubuntu/authd/blob/main/internal/testsdetection/ if you really need to put some part in production code to assert quickly that this is not allowed: https://github.com/ubuntu/authd/blob/main/internal/services/permissions/testutils.go.

Do not then hesitate to import, like from the `_linux.go` file to attach it to the ipConfig or wslSystem for instance (this, combined with the previous advice on how to attach it to those struct should give a good start!)

Finally, there is the need for integration tests (but it seems, we are not there yet!),. We can reuse then the import in the finale binary, like [https://github.com/ubuntu/authd/blob/main/internal/testsdetection/ and ](https://github.com/ubuntu/authd/blob/main/cmd/authd/integrationtests.go.)